common/channelconfig/util.go

```diff
@@ -241,51 +242,72 @@ func ACLValues(acls map[string]string) *StandardConfigValue {
 
 // ValidateCapabilities validates whether the peer can meet the capabilities requirement in the given config block
 func ValidateCapabilities(block *cb.Block, bccsp bccsp.BCCSP) error {
+	cc, err := extractChannelConfig(block, bccsp)
+	if err != nil {
+		return err
+	}
+	// Check the channel top-level capabilities
+	if err := cc.Capabilities().Supported(); err != nil {
+		return err
+	}
+
+	// Check the application capabilities
+	return cc.ApplicationConfig().Capabilities().Supported()
+}
+
+// ExtractMSPIDsForApplicationOrgs extracts MSPIDs for application organizations
+func ExtractMSPIDsForApplicationOrgs(block *cb.Block, bccsp bccsp.BCCSP) ([]string, error) {
+	cc, err := extractChannelConfig(block, factory.GetDefault())
+	if err != nil {
+		return nil, err
+	}
+
+	if cc.ApplicationConfig() == nil {
+		return nil, errors.Errorf("could not get application config for channel")
+	}
+	orgs := cc.ApplicationConfig().Organizations()
+	mspids := make([]string, 0, len(orgs))
+	for _, org := range orgs {
+		mspids = append(mspids, org.MSPID())
+	}
+	return mspids, nil
+}
+
+func extractChannelConfig(block *cb.Block, bccsp bccsp.BCCSP) (*ChannelConfig, error) {
 	envelopeConfig, err := protoutil.ExtractEnvelope(block, 0)
 	if err != nil {
-		return errors.Errorf("failed to %s", err)
+		return nil, errors.Errorf("failed to %s", err)
 	}
 
 	configEnv := &cb.ConfigEnvelope{}
 	_, err = protoutil.UnmarshalEnvelopeOfType(envelopeConfig, cb.HeaderType_CONFIG, configEnv)
 	if err != nil {
-		return errors.Errorf("malformed configuration envelope: %s", err)
+		return nil, errors.Errorf("malformed configuration envelope: %s", err)
 	}
 
 	if configEnv.Config == nil {
-		return errors.New("nil config envelope Config")
+		return nil, errors.New("nil config envelope Config")
 	}
 
 	if configEnv.Config.ChannelGroup == nil {
-		return errors.New("no channel configuration was found in the config block")
+		return nil, errors.New("no channel configuration was found in the config block")
 	}
 
 	if configEnv.Config.ChannelGroup.Groups == nil {
-		return errors.New("no channel configuration groups are available")
+		return nil, errors.New("no channel configuration groups are available")
 	}
 
 	_, exists := configEnv.Config.ChannelGroup.Groups[ApplicationGroupKey]
 	if !exists {
-		return errors.Errorf("invalid configuration block, missing %s "+
+		return nil, errors.Errorf("invalid configuration block, missing %s "+
 			"configuration group", ApplicationGroupKey)
```

Comment:
We are already using Errorf, can we remove the concatenation

Reply:
fwiw we may want to change this to `fmt.Errorf` simply for the reason that `github.com/pkg/errors` will no longer support new features and the shift is towards using golang's `errors` package and delegating formatted errors to `fmt.Errorf`

Reply:
@btl5037 thanks for pointing it out. This is existing code, but should be fixed since we are changing it anyway.
@caod123 thanks for pointing it out. In this PR, I prefer to keep errors.Errorf to be consistent with the rest of code. If it is decided to change to fmt.Errorf, we may want to have a separate PR for the change.